### ISSUE TITLE
Decoupling runtime frameworks (angularJs) and e2e testing framewoks (protractor)

### ISSUE THREAD

**Issue by alfonso-presa** [angular/protractor#2246]
## TL;DR 
This issue proposes to build an spec to enable decoupling between testing frameworks and the runtime frameworks, libraries and applications.

## Intro
I've been working for more than a year with protractor and angular and I love both tools, but I've come to some issues when trying to make protractor work when building some modules and applications for angular (specially when playing with the router and when doing lazyloading).

In this time I've learned that in order to make protractor work (wait natively for everything to be ready in the page without aditional waits for this and for that in the test spec) you should use angular API "correctly" in your development. The best pattern to do so I've come with is using $browser.defer(...).cancel(...) (please if there's a better one please let me know).

Any way I think it's weird that protractor needs angular so badly to work. I understand that it provides some added value like model or directive based selectors, but IMHO the "waitForAngular" service should be something more abstract like "waitForBrowserToBeReady".

## Proposal

I think it will be better for the OpenSource community if there was an abstraction layer (hopefully an spec) between the runtime libraries like angular and the testing frameworks like protractor. This will have the following advantages:
- Testing frameworks will not be coupled to specific framework or versions.
- Runtime frameworks will be testable by any testing framework, present or future.
- Last but not least, applications and libraries will have a mechanism to create aditional testable asynchronous services outside of the frameworks (like websockets, webworkers, lazyloading, custom rendering algorithms, etc).

## Sample implementation
I've made a first reference implementation for this future spec for you to check the viability of the proposal: https://github.com/alfonso-presa/testability.js

**Comment by sjelin** [comment]
Interesting idea, since non-angular users are increasing using protractor.  Our user base is still largely angular though, and I don't think anyone on the team has enough bandwidth to undertake this kind of project.  We are currently trying to make Protractor more adaptable to different apps though things like [plugins](https://github.com/angular/protractor/blob/master/docs/plugins.md), and I think in the near-term that's about all we'll be able to do.

**Comment by alfonso-presa** [comment]
Thank you for the quick reply. I was thinking of doing this myself via PR, but at this point is even better to handle it through a plugin as you say before polluting protractor with something that's not proven to success.

Also note the aim of this is not only frameworks other than angular but also enable angular applications to do some stuff or use libraries not currently under angular control without having to monkey patch their e2e tests. I've been in this situation a couple of times in real projects and I got tired of listening complains about protractor not working when the problem is that there's not a clear API to declare that the current state is not consistent.

If you want an example of this let's say you have a websocket communication in an angular application, you click a button an then a message is sent to the backend...the application is the only one who knows if the current page state is testable or if it has to wait for a message to arrive from the server. People will expect protractor to magically know that for them and when it doesn't work they start complaining an playing with things like implicit wait which makes the tests last for ever (due to assertions of elements not to be present for example) and also makes the tests flakey which makes the "testing experience" even worse. I don't think this situation is fair with a really nice piece of software like protractor :-).

I'm also planning to make some PR for the angular project to document the way I think this can be done right now by using $browser.defer().cancel().

Thank you!

Keep up the good work guys!

**Comment by sjelin** [comment]
We do have a [pending PR](https://github.com/angular/protractor/pull/2104) which allows you more flexibility in what you wait for, and I think it addresses your websocket usecase

**Comment by alfonso-presa** [comment]
Woohoo. Yes that look's a good step forward :-). I guess it will allow developing a plugin that can wait for some condition in the browser. It will be a very good way to integrate my spec's plugin!

Thanks!

**Comment by alfonso-presa** [comment]
This is what I have so far. Good news is it is working with and without angular in scope :-).
https://github.com/alfonso-presa/protractor-testability-plugin

Testing non angular async stuff is as easy as including the plugin and doing:
```js
if(testability) { testability.wait.for(promise);}
```
where ever you have an async resolved promise (q style). You can also make it based in count but it's less safe.

Bad news is that I've had to provide an angular stub object in case angular is not present:
https://github.com/alfonso-presa/protractor-testability-plugin/blob/master/lib/protractor-testability-plugin.js#L14

Also I've used something I believe is protractor's private API (is there something public I can use for this without duplicating protractor's code?):
https://github.com/alfonso-presa/protractor-testability-plugin/blob/master/lib/protractor-testability-plugin.js#L43

Other than that it was really straight forward to implement. Great work with #2104.

Any recommendations and feedback will be very welcome :-)

Thanks!

**Comment by sjelin** [comment]
You couldn't use the public `browser.executeAsyncScript`?

**Comment by alfonso-presa** [comment]
Ops... sounds like that could do it :-P. Thanks!

**Comment by alfonso-presa** [comment]
It worked perfectly with executeAsyncScript. Thank you again.

I guess that when #2104 gets released, plus my plugin, the spirit of this issue will be functionally covered :-).

**Comment by alfonso-presa** [comment]
Now that 2.2.0 has been released, I've tested my plugin with it and is working like a charm.

Do you think it will be ok if I PR with the addition of "community plugins" in `docs/plugins.md` and include there mine with a little description of what it does? I think my plugin might be very useful for people developing the scenarios stated in the beginning of this issue (non-angular applications and asynchronous stuff outside angular's control, for example web-socket communication).
